### PR TITLE
Make the async rate limiter implementation a bit more robust.

### PR DESCRIPTION
**Pull request by jacob314**

### CODE REVIEW

File: testSrc/unit/io/flutter/utils/AsyncRateLimiterTest.java [hunk 1]
```diff
@@ -111,11 +108,13 @@ public void rateLimited() {
     checkLog("EVENT", "EVENT", "EVENT", "EVENT", "DONE");
 
     // First event should occur immediately so don't count it.
-    final double requestsPerSecond = (expectedEvents - 1) / (delta * 0.001);
+    // The last event just fired so add in MS_PER_EVENT to factor in that it
+    // will be that long before the next event fires.
+    final double requestsPerSecond = (numEvents - 1) / ((delta + MS_PER_EVENT) * 0.001);
 
     // Use an epsilon of 10% for comparisons to TEST_FRAMES_PER_SECOND.
```
Comment (stevemessick) [inline testSrc/unit/io/flutter/utils/AsyncRateLimiterTest.java]: Is this comment still valid?

File: src/io/flutter/utils/AsyncRateLimiter.java [hunk 2]
```diff
@@ -41,57 +49,81 @@ public AsyncRateLimiter(double framesPerSecond, Computable<CompletableFuture<?>>
...
+      if (numRequestsNotYetStarted == 0) {
+        numRequestsPending++;
+      }
     }
+    if (scheduleRequestImmediately) {
+      scheduleRequestHelper();
+    }
+  }
 
+  // THis method may be called on any thread.
```
Comment (stevemessick) [inline src/io/flutter/utils/AsyncRateLimiter.java]: "This"

File: src/io/flutter/utils/AsyncRateLimiter.java [hunk 3]
```diff
@@ -41,57 +49,81 @@ public AsyncRateLimiter(double framesPerSecond, Computable<CompletableFuture<?>>
...
+  private void performRequestOnUiThread() {
+    final Runnable doRun = () -> {
+      if (requestScheduler.isDisposed()) {
+        return;
+      }
       performRequest();
+    };
+    final Application app = ApplicationManager.getApplication();
+    if (app == null || app.isUnitTestMode()) {
+      // This case existing to support unittesting.
```
Comment (stevemessick) [inline src/io/flutter/utils/AsyncRateLimiter.java]: "exists"

Comment (jacob314) [reply src/io/flutter/utils/AsyncRateLimiter.java]: done

File: src/io/flutter/utils/AsyncRateLimiter.java (hunk 2 above)
Comment (jacob314) [reply src/io/flutter/utils/AsyncRateLimiter.java]: Done

File: testSrc/unit/io/flutter/utils/AsyncRateLimiterTest.java (hunk 1 above)
Comment (jacob314) [reply testSrc/unit/io/flutter/utils/AsyncRateLimiterTest.java]: Good point. Fixed the bug in the test that required it. Removed.

**Pull request merged**